src/vmm/src/controller.rs

```diff
@@ -225,4 +227,243 @@ impl VmmController {
 
         Ok(())
     }
+
+    /// Retrieves the KVM dirty bitmap for each of the guest's memory regions.
+    pub fn get_dirty_bitmap(
+        &self,
+    ) -> std::result::Result<HashMap<usize, Vec<u64>>, VmmActionError> {
+        let mut bitmap: HashMap<usize, Vec<u64>> = HashMap::new();
```

Comment:
There's no reason for slot indices to be contiguous and some users might want to leverage that in the future, so I wouldn't simplify it to a plain `Vec` (a `Vec<Option<...>>` would work but it would make the code really ugly). The `Vec<(slot, Vec<u64>)>` solution gets rid of the hashing and keeps the map-like correspondence, but it increases per-slot access time for the consumer, and I'd rather avoid that at the expense of hashing.